iceoryx2-bb/elementary/src/relocatable.rs

```diff
@@ -0,0 +1,113 @@
+///    `#[repr(C)]`
+///  * The type does not have references or pointer members. It shall not use pointers to manage
+///    its internal structure.
+///    Examples:
+///       * A list must be implemented using indices to structure itself.
+pub unsafe trait Relocatable {}
+
+unsafe impl Relocatable for usize {}
+unsafe impl Relocatable for u8 {}
+unsafe impl Relocatable for u16 {}
+unsafe impl Relocatable for u32 {}
+unsafe impl Relocatable for u64 {}
+unsafe impl Relocatable for u128 {}
+
+unsafe impl Relocatable for isize {}
+unsafe impl Relocatable for i8 {}
+unsafe impl Relocatable for i16 {}
+unsafe impl Relocatable for i32 {}
+unsafe impl Relocatable for i64 {}
+unsafe impl Relocatable for i128 {}
+
+unsafe impl Relocatable for f32 {}
+unsafe impl Relocatable for f64 {}
+
+unsafe impl Relocatable for char {}
+unsafe impl Relocatable for bool {}
+
+unsafe impl Relocatable for IoxAtomicUsize {}
+unsafe impl Relocatable for IoxAtomicU8 {}
+unsafe impl Relocatable for IoxAtomicU16 {}
+unsafe impl Relocatable for IoxAtomicU32 {}
+unsafe impl Relocatable for IoxAtomicU64 {}
+
+unsafe impl Relocatable for IoxAtomicIsize {}
+unsafe impl Relocatable for IoxAtomicI8 {}
+unsafe impl Relocatable for IoxAtomicI16 {}
+unsafe impl Relocatable for IoxAtomicI32 {}
+unsafe impl Relocatable for IoxAtomicI64 {}
+
+unsafe impl<T: Relocatable> Relocatable for [T] {}
+unsafe impl<T: Relocatable, const N: usize> Relocatable for [T; N] {}
+unsafe impl<T: Relocatable> Relocatable for Option<T> {}
+unsafe impl<T: Relocatable, E: Relocatable> Relocatable for Result<T, E> {}
+unsafe impl<T: Relocatable> Relocatable for core::mem::MaybeUninit<T> {}
+unsafe impl<T: Relocatable> Relocatable for core::cell::UnsafeCell<T> {}
+
+unsafe impl<T1: Relocatable, T2: Relocatable> Relocatable for (T1, T2) {}
+unsafe impl<T1: Relocatable, T2: Relocatable, T3: Relocatable> Relocatable for (T1, T2, T3) {}
+unsafe impl<T1: Relocatable, T2: Relocatable, T3: Relocatable, T4: Relocatable> Relocatable
+    for (T1, T2, T3, T4)
+{
+}
+unsafe impl<T1: Relocatable, T2: Relocatable, T3: Relocatable, T4: Relocatable, T5: Relocatable>
+    Relocatable for (T1, T2, T3, T4, T5)
+{
+}
+unsafe impl<
+        T1: Relocatable,
+        T2: Relocatable,
+        T3: Relocatable,
+        T4: Relocatable,
+        T5: Relocatable,
+        T6: Relocatable,
+    > Relocatable for (T1, T2, T3, T4, T5, T6)
+{
+}
+unsafe impl<
+        T1: Relocatable,
+        T2: Relocatable,
+        T3: Relocatable,
+        T4: Relocatable,
+        T5: Relocatable,
+        T6: Relocatable,
+        T7: Relocatable,
+    > Relocatable for (T1, T2, T3, T4, T5, T6, T7)
+{
+}
+unsafe impl<
+        T1: Relocatable,
+        T2: Relocatable,
+        T3: Relocatable,
+        T4: Relocatable,
+        T5: Relocatable,
+        T6: Relocatable,
+        T7: Relocatable,
+        T8: Relocatable,
+    > Relocatable for (T1, T2, T3, T4, T5, T6, T7, T8)
+{
+}
```

Comment:
I've been looking for a way to generalize this for tuples with any number of elements, but https://doc.rust-lang.org/std/primitive.tuple.html#impl-Debug-for-(T,) makes me doubt that it's possible :thinking:

Reply:
Unfortunately, it's not yet possible. You are in the happy siuation that it works for arrays in a generic way. There was a time where many built-in traits were only implemented for arrays up to 32 elements and then a few more with element sizes of power if two. Oh, these good old times 😅